File: apps/navigation/backed-js/src/routes/navigation.js
import multer from 'multer'
import { getRoadStatus, findRoute } from '../services/routing.js'
import { importGpkgStream } from '../services/gpkgImport.js'

const upload = multer({ storage: multer.memoryStorage() })

export const roadStatusHandler = async (req, res) => {
    const result = await getRoadStatus()
    res.json({ code: '200', data: result })
}

export const routeHandler = async (req, res) => {
    const { start_lon, start_lat, end_lon, end_lat } = req.query
    if (!start_lon || !start_lat || !end_lon || !end_lat) {
        res.json({ code: '500', data: '缺少 start_lon/start_lat/end_lon/end_lat 参数' })
        return
    }

    const result = await findRoute(
        parseFloat(start_lon),
        parseFloat(start_lat),
        parseFloat(end_lon),
        parseFloat(end_lat),
    )

    if (result.found) {
        res.json({ code: '200', data: result })
    } else {
        res.json({ code: '500', data: result.msg })
    }
}

export const importGpkgHandler = [
    upload.single('file'),
    async (req, res) => {
        if (!req.file) {
            res.json({ code: '500', data: '请上传文件' })
            return
        }

        if (!req.file.originalname.toLowerCase().endsWith('.gpkg')) {
            res.json({ code: '500', data: '请上传 .gpkg 文件' })
            return
        }

        res.setHeader('Content-Type', 'text/event-stream')
        res.setHeader('Cache-Control', 'no-cache')
        res.setHeader('X-Accel-Buffering', 'no')
        res.flushHeaders()

        await importGpkgStream(req.file.buffer, req.file.originalname, res)
    },
]

File: apps/navigation/backed/app/api/navigation.py
"""导航相关 API 路由"""
import logging

from fastapi import APIRouter, UploadFile, File, Query
from fastapi.responses import StreamingResponse

from app.services.gpkg_import import import_gpkg_to_postgis
from app.services.routing import get_road_status, find_route
from app.utils.response import success, error

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api")


@router.get("/road-status")
async def road_status():
    result = get_road_status()
    return success(result)


@router.post("/import-gpkg")
async def import_gpkg(file: UploadFile = File(...)):
    if not file.filename or not file.filename.lower().endswith(".gpkg"):
        return error("请上传 .gpkg 文件")

    gpkg_bytes = await file.read()
    if not gpkg_bytes:
        return error("文件为空")

    return StreamingResponse(
        import_gpkg_to_postgis(gpkg_bytes, file.filename),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",
        },
    )


@router.get("/route")
async def route(
    start_lon: float = Query(...),
    start_lat: float = Query(...),
    end_lon: float = Query(...),
    end_lat: float = Query(...),
):
    result = find_route(start_lon, start_lat, end_lon, end_lat)
    if result.get("found"):
        return success(result)
    return error(result.get("msg", "未找到路径"))
